app/Community/Actions/BuildActivePlayersAction.php

```diff
@@ -0,0 +1,175 @@
+        $offset = ($page - 1) * $perPage;
+        $paginatedPlayers = array_slice($activePlayersList, $offset, $perPage);
+
+        $hydratedPlayers = $this->hydrateResults($paginatedPlayers);
+
+        $paginator = new LengthAwarePaginator(
+            items: $hydratedPlayers,
+            total: count($activePlayersList),
+            perPage: $perPage,
+            currentPage: $page,
+        );
+
+        return PaginatedData::fromLengthAwarePaginator($paginator, unfilteredTotal: $unfilteredTotal);
+    }
+
+    /**
+     * @param array<array{
+     *   user_id: int,
+     *   game_id: int,
+     *   username: string,
+     *   display_name: string,
+     *   rich_presence: string,
+     *   game_title: string,
+     * }> $players
+     *
+     * @return array<array{
+     *   user_id: int,
+     *   game_id: int,
+     *   username: string,
+     *   display_name: string,
+     *   rich_presence: string,
+     *   game_title: string,
+     * }>
+     */
+    private function filterByGameIds(array $players, ?array $gameIds = null): array
+    {
+        if (!$gameIds) {
+            return $players;
+        }
+
+        return array_filter(
+            $players,
+            fn ($player) => in_array($player['game_id'], $gameIds, strict: true)
+        );
+    }
+
+    /**
+     * @param array<array{
+     *   user_id: int,
+     *   game_id: int,
+     *   username: string,
+     *   display_name: string,
+     *   rich_presence: string,
+     *   game_title: string,
+     * }> $players
+     *
+     * @return array<array{
+     *   user_id: int,
+     *   game_id: int,
+     *   username: string,
+     *   display_name: string,
+     *   rich_presence: string,
+     *   game_title: string,
+     * }>
+     */
+    private function filterBySearch(array $players, ?string $search): array
+    {
+        if (!$search) {
+            return $players;
+        }
+
+        $search = Str::lower(trim($search));
+
+        // We should also support logical OR searches, for example:
+        // "developing|inspecting".
+        $searchTerms = array_map(
+            fn (string $term) => Str::lower(trim($term)),
+            explode('|', $search)
+        );
+
+        // Filter out any empty terms.
+        $searchTerms = array_filter($searchTerms);
+        if (empty($searchTerms)) {
+            return $players;
+        }
+
+        return array_filter(
+            $players,
+            function ($player) use ($searchTerms) {
+                // For each search term, check if it exists in any of the fields.
+                foreach ($searchTerms as $term) {
+                    if (
+                        str_contains(Str::lower($player['username']), $term)
+                        || str_contains(Str::lower($player['display_name']), $term)
+                        || str_contains(Str::lower($player['rich_presence']), $term)
+                        || str_contains(Str::lower($player['game_title']), $term)
```

Comment:
I can't help but wonder if there's a way to keep the lowercased values between iterations of the `$searchTerms` loop.

Google suggests using `mb_stripos` for case-insensitive matching, but it's probably a lot more expensive than normalizing to lowercase and doing a string compare.

Reply:
This was mostly copied from `ActivePlayersService` (which is currently using `stripos()`), but I agree it can be heavily optimized.

In latest, I concatenate all the searchable fields into a single string and then lowercase it one time. Then I do `str_contains($searchableText, $term)`.